oml/functional/metrics.py

```diff
@@ -220,42 +216,209 @@ def calculate_accuracy_on_triplets(embeddings: torch.Tensor, reduce_mean: bool =
         return acc
 
 
+def calc_cmc(gt_tops: torch.Tensor, top_k: Tuple[int, ...]) -> List[torch.Tensor]:
+    """
+    Function to compute Cumulative Matching Characteristics (CMC) for each sample.
+
+    ``cmc@k`` is 1 if ``top_k`` distances contain an element from the same class, and 0 otherwise.
+
+    Args:
+        gt_tops: ground truth of the ``max(top_k)`` distances.
```

Comment:
Decided to describe `gt_tops`

Reply:
I think let's keep the function as it is but use my comment above for a doc string

Reply:
@DaloroAT, @AlekseySh

Actually, I suggested the changes in the `calc_*` below

> `gt_top` is derived from `mask_gt`, so maybe it will be better to change the `calc_*` signatures so that they look like

```python
def calc_cmc(mask_gt: Tensor, top_indexes: torch.Tensor, top_k: Tuple[int, ...]) -> Tensor:
    """
    Args:
        mask_gt: ...
        top_indexes: Tensor of indices of the smallest k elements from the query-to-gallery distance matrix.
        ...
    """
```

> But then the following lines will be done in three functions:

```python
    ii_arange = torch.arange(query_sz).unsqueeze(-1).expand(query_sz, max_k)
    gt_tops = mask_gt[ii_arange, ii_top_k]
```

> Of course these lines should be placed in a separate function say `_extract_gt_tops`, but the computation still will take some time.

What do you think about it? Is it ok to get `gt_tops` inside each `calc_*` function? Maybe some sort of a caching could be used here to reuse `gt_tops`, say simple `lru_cache` with `maxsize=1`. (I don't know whether it is safe to use `torch.Tensor` as a hashable variable, are there any drawbacks of this idea?)

If so, maybe we can also cache this `torch.topk(distances, k=max_k, largest=False)`, and change the `calc_*` signature to

```python
def calc_cmc(mask_gt: Tensor, distances: torch.Tensor, top_k: Tuple[int, ...]) -> Tensor:
```